src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+  // a "trap" mover
+  EntryMover(const parent_t&, from_t&, to_t& dst, unsigned) {
+    assert(0);
+  }
+  void move_from(unsigned, unsigned, unsigned) {
+    assert(0);
+  }
+  delta_t get_delta() {
+    return delta_t::nop();
+  }
+};
+
+// lower the layout, for instance, from L0 to L1, no reference oid is used
+template<class parent_t,
+         class from_t,
+         class to_t>
+class EntryMover<parent_t,
+                 from_t,
+                 to_t,
+                 std::enable_if_t<from_t::node_n < to_t::node_n>>
+{
+public:
+  EntryMover(const parent_t&, from_t& src, to_t& dst, unsigned)
+    : src{src}, dst{dst}
+  {}
+  void move_from(unsigned src_first, unsigned dst_first, unsigned n)
+  {
+    ceph::bufferptr keys_buf{n * sizeof(to_t::key_t)};
+    ceph::bufferptr cells_buf;
+    auto dst_keys = reinterpret_cast<typename to_t::key_t*>(keys_buf.c_str());
+    if constexpr (to_t::item_in_key) {
+      for (unsigned i = 0; i < n; i++) {
+        const auto& [prefix, suffix] = src.key_at(src_first + i);
+        dst_keys[i].set(suffix, src.item_at(prefix));
+      }
+    } else {
+      // copy keys
+      uint16_t src_offset = src_first > 0 ? src.keys[src_first - 1].offset : 0;
+      uint16_t dst_offset = dst_first > 0 ? dst.keys[dst_first - 1].offset : 0;
+      for (unsigned i = 0; i < n; i++) {
+        auto& src_key = src.keys[src_first + i];
+        uint16_t offset = src_key.offset - src_offset + dst_offset;
+        dst_keys[i].set(src_key, offset);
+      }
+      // copy cells in bulk, yay!
+      auto src_end = src.keys[src_first + n - 1].offset;
+      uint16_t total_cell_size = src_end - src_offset;
+      cells_buf = ceph::bufferptr{total_cell_size};
+      cells_buf.copy_in(0, total_cell_size, src.from_end(src_end));
+    }
+    if (dst_first == dst.count) {
+      dst_delta = delta_t::insert_back(keys_buf, cells_buf);
+    } else {
+      dst_delta = delta_t::insert_front(keys_buf, cells_buf);
+    }
+    if (src_first > 0 && src_first + n == src.count) {
+      src_delta = delta_t::trim_right(src_first);
+    } else if (src_first == 0 && n < src.count) {
+      src_delta = delta_t::shift_left(n);
+    } else if (src_first == 0 && n == src.count) {
+      // the caller will retire the src extent
+    } else {
+      // grab in the middle?
+      assert(0);
+    }
+  }
+
+  delta_t from_delta() {
+    return std::move(src_delta);
+  }
+  delta_t to_delta() {
+    return std::move(dst_delta);
+  }
+private:
+  const from_t& src;
+  const to_t& dst;
+  delta_t dst_delta;
+  delta_t src_delta;
+};
+
+// lift the layout, for instance, from L2 to L0, need a reference oid
+template<class parent_t,
+         class from_t,
+         class to_t>
+class EntryMover<parent_t, from_t, to_t,
+                 std::enable_if_t<(from_t::node_n > to_t::node_n)>>
+{
+public:
+  EntryMover(const parent_t& parent, from_t& src, to_t& dst, unsigned from_slot)
+    : src{src}, dst{dst}, ref_oid{parent->update_oid_with_slot(from_slot, {})}
```

Comment:
I guess `from_slot` is the parent index with a logical address pointing to the child node `src`, right?

Could you elaborate (could be somewhere in the comment) what are the implications between the parent indexes and the children nodes in your current implementation?

Specifically:
```
P[L0: ..., i, i+1, ...]
          /      \
Ci[L1: 0, ..., j] Ci+1[L0: 0, ..., k]
```
If I remembered correctly from our conversations last time, you said that:
* P[i] <= Ci[0] && Ci[j] < P[i+1] && P[i+1] <= Ci+1[0];
* The index P[i] doesn't necessarily need to match/be equal to Ci[0];
* The index P[i+1] doesn't necessarily need to match/be equal to Ci+1[0];

Reply:
sure, will document this in the comment to outline the difference between my implementation and the classical b+ tree.

Reply:
@cyx1231st please note, the stuff you outlined above is possible **policies** i am imaging. this changeset does not imply them in any way.